chore: add visitors

maastrich opened the pull request:
# what does it do?

Add visitors for remaining nodes types with tests for most of them 

# why?
I was planning do make a similar package but you seem to have handled most of the work and that I could handle some of the rest

maastrich commented:
Hello @morganney I was thinking that you could use some help, tell me if it's ok ;)

maastrich commented:
Almost done, but a few things to take care about aggregate types 
I'll take care of them in the morning, it's getting late in paris

morganney commented:
bonne nuit.

maastrich commented:
hello @morganney, I've updated the PR with some tests, do you think this kind of tests works for you ?

morganney commented:
I think it is a solid approach, I like it. Thanks. I've got a couple questions/comments though if you don't mind.

morganney commented on src/types.ts:
```diff
@@ -110,6 +108,12 @@ export type AnyNode =
   | swc.MetaProperty
   | swc.MethodProperty
   | swc.Module
+  | swc.ExportNamespaceSpecifier
+  | swc.ExportDefaultSpecifier
+  | swc.NamedExportSpecifier
+  | swc.NamedImportSpecifier
+  | swc.ImportDefaultSpecifier
+  | swc.ImportNamespaceSpecifier
```
I believe these are all already there. For example `swc.ExportNamespaceSpecifier` is also on line 73. Double check that these aren't duplicates.

maastrich replied:
I've added `@typescript-eslint/no-duplicate-type-constituents` in order to avoid this

morganney commented on test/baseVisitors.test.ts:
```diff
@@ -0,0 +1,333 @@
+  if (options.debug) {
+    console.log(JSON.stringify(ast, null, 2))
+  }
+  simple(ast, {}, baseVisitor)
```
I missed last time that `walk.ts` should add `Node` to the [type import](https://github.com/morganney/swc-walk/blob/main/src/walk.ts#L4) so that the underlying type is from an SWC Node and not a DOM Node.

morganney commented on src/types.ts:
```diff
@@ -19,9 +19,7 @@ export type AggregateType = {
   ClassMember: swc.ClassMember
   Declaration: swc.Declaration
   DefaultDecl: swc.DefaultDecl
-  ExportSpecifier: swc.ExportSpecifier
```
What was the reason for removing these from the `AggregateType`? Same for `swc.ImportSpecifier` below?

maastrich replied:
I was having conflict between `ExportSpecifier` the aggregate type and `NamedExportSpecifier` whose type is `"ExportSpecifier"`
I'm still planning on restoring theme

morganney commented on test/baseVisitors.test.ts:
```diff
@@ -0,0 +1,333 @@
+  },
+  { type: 'TsAsExpression', code: 'a as b', mock: mock.method(baseVisitor, 'TsAsExpression') },
+  {
+    type: 'TsCallSignatureDeclaration',
+    code: 'type foo = () => void',
+    mock: mock.method(baseVisitor, 'TsCallSignatureDeclaration'),
```
Not entirely sure why yet, but when I ran the tests locally this one failed:

```
✖ should call TsCallSignatureDeclaration (157.277577ms)
  AssertionError [ERR_ASSERTION]: Expected TsCallSignatureDeclaration to be called 1 times, but was called 0 times.
      at Object.toHaveBeenCalledTimes (/Users/money.ctr/morganney/swc-walk/test/baseVisitors.test.ts:326:14)
      at TestContext.<anonymous> (/Users/money.ctr/morganney/swc-walk/test/baseVisitors.test.ts:310:28)
      at Test.runInAsyncScope (node:async_hooks:206:9)
      at Test.run (node:internal/test_runner/test:639:25)
      at Test.processPendingSubtests (node:internal/test_runner/test:382:18)
      at Test.postRun (node:internal/test_runner/test:730:19)
      at Test.run (node:internal/test_runner/test:688:12)
      at async Test.processPendingSubtests (node:internal/test_runner/test:382:7) {
    generatedMessage: false,
    code: 'ERR_ASSERTION',
    actual: 0,
    expected: 1,
    operator: '=='
  }
```

maastrich replied:
yes, gotta find the correct test to run

morganney replied:
Ah, I see. Yeah I had been trying to find the best way to match a node to the source code. I considered using `import { printSync } from '@swc/core'` and then running an AST node through that function to generate the corresponding source, but at that point might as well just run the constructed node through the test. I think I'd still prefer to run the tests from source code that is parsed, rather than a direct AST node, but technically all we need are AST nodes not the corresponding source code.

Did you have a preference on starting with a piece of source code vs. starting with an AST node?

maastrich replied:
The problem with the AST node is to find, the correct parents nodes
for the test whose are ready yet, I've either relied on Github Copilot or astexplorer.net

morganney commented on test/baseVisitors.test.ts:
```diff
@@ -0,0 +1,333 @@
+  })
+}
+
+function given(options: (typeof tests)[number]) {
```
This is a nitpick, and I know that function declarations and their body are hoisted to the top of the module scope, but I would prefer that this was not used before it was defined.

I generally enable [`no-use-before-define`](https://typescript-eslint.io/rules/no-use-before-define/). Feel free to add that lint rule.

maastrich replied:
As u want, I usually do like that as a DX good practice, those functions are the least important things in the file, they are "blackboxes" that we do not need to see while opening the file
This way, when I open the file, I directly have my tests

morganney replied:
I do understand the practicality of keeping the relevant parts first. This is what they refer to as bike shedding :smile: 

You can also move these functions to some `util.ts` file within `test/`.

morganney commented on test/baseVisitors.test.ts:
```diff
@@ -0,0 +1,333 @@
+  return options
+}
+
+function expect(options: (typeof tests)[number]) {
```
Same, I would prefer if we did not rely on hoisting.

morganney commented on src/baseVisitor.ts:
```diff
@@ -90,199 +89,346 @@ export class BaseVisitor<T> implements Required<RecursiveVisitors<T>> {
+    for (const decorator of n.decorators ?? []) {
+      cb(decorator, st)
+    }
+    cb(n.identifier, st)
```
This is simply a style preference of mine, but you can fix all of those `Expected blank line before this statement` lint errors by running `npm run lint:fix`.

morganney commented:
@maastrich I've created branch [pr/2-suggested-changes](https://github.com/morganney/swc-walk/tree/pr/2-suggested-changes) if you wanted to rebase this work off of that. Either that or work from my comments, up to you. I appreciate your help. I'll be able to continue collaborating after work.

maastrich commented:
Hey @morganney, I've made some pretty good advancements, coverage is starting to look good and I've patched some of the flaws

Next step is to reset aggregated types and find a workaround about the name conflicts

I've settled on a few tests skips as I do not know how to trigger these nodes visitors

about the tests utils, I moved them in a utils file as you proposed 

PS: I've made another PR about AggregatedTypes -> https://github.com/morganney/swc-walk/pull/5

maastrich committed:
Squash chore/add-visitors (merge-base 72c04a66c10c5605f76787d6541623c41715ba2e, previous tip a3c417baa448fdc27efb530b0e7f450a8108ba1c)

commit a3c417baa448fdc27efb530b0e7f450a8108ba1c
Author: maastrich <pinsaultm@gmail.com>
Date:   Wed Jun 12 20:54:32 2024 +0200

    chore: Remove console.error in AssignmentProperty method

commit 281106dd7e267b55dfd26377da50eb3a317a02da
Author: maastrich <pinsaultm@gmail.com>
Date:   Wed Jun 12 20:46:56 2024 +0200

    chore: Enable decorators in parseSync options

commit b59b57a1b166dfc61865d235a7339070db1c7c39
Author: maastrich <pinsaultm@gmail.com>
Date:   Wed Jun 12 20:07:49 2024 +0200

    feat: Add missing callbacks for function properties in BaseVisitor

commit 631e4f262f7b39635ca57721be75847e9047c2cd
Author: maastrich <pinsaultm@gmail.com>
Date:   Wed Jun 12 18:13:35 2024 +0200

    chore: Update npm dependency to latest stable version

commit 54ee886bb3195381bb719d4eacc02572a335eb01
Author: maastrich <pinsaultm@gmail.com>
Date:   Wed Jun 12 16:13:18 2024 +0200

    feat: Add missing callbacks for function properties in BaseVisitor

commit e626433190244c0b861d76965e2993090c1ec99a
Author: maastrich <pinsaultm@gmail.com>
Date:   Wed Jun 12 16:05:10 2024 +0200

    chore: Update npm dependency to latest stable version

commit fe095156ac08449248831c554445acd3ccb136a1
Author: maastrich <pinsaultm@gmail.com>
Date:   Tue Jun 11 19:27:19 2024 +0200

    feat: Add missing callbacks for function properties in BaseVisitor

commit 5c21cbc82bb9eede916b5e9a1e0e4156c230c9b5
Author: maastrich <pinsaultm@gmail.com>
Date:   Tue Jun 11 17:31:16 2024 +0200

    feat: Add missing callbacks for function properties in BaseVisitor

commit e86fc0c4d5d33d1811c7aeac8764a0c7551867a3
Author: maastrich <pinsaultm@gmail.com>
Date:   Tue Jun 11 02:28:27 2024 +0200

    feat: Implement all node walkers method

commit dab2fd86d8fdfefec5f924341150f027dd833509
Author: maastrich <pinsaultm@gmail.com>
Date:   Tue Jun 11 01:37:41 2024 +0200

    chore: Implement all node walkers method

commit 8f967ab22676a9a868f403e7a5d89ae9b02a2d6c
Author: Mathis Pinsault <pinsaultm@gmail.com>
Date:   Tue Jun 11 01:56:53 2024 +0200

    chore: add all node walkers method as no implemented (#3)

    * chore: add all node walkers method as no implemented

    * refactor: remove unused import in walk.ts

commit e88099c2f8fcff9807511591d5890038e9e98467
Author: Morgan Ney <morganney@gmail.com>
Date:   Mon Jun 10 18:39:17 2024 -0500

    feat: add more nodes. (#4)

    * chore: add style preferences as lint rules.

maastrich committed:
chore: Update npm dependency to latest stable version

maastrich commented:
@morganney I believe that I'm happy with these changes
I've opened [a PR](https://github.com/swc-project/swc/pull/9069) on `swc` to add a missing type (forcing me to add a `@ts-expect-error`) i'm hopping that it will be merged en published soon

morganney commented:
@maastrich mind rebasing with `main`? I made some changes. Specifically, I wasn't sure why `tsconfig.meta.json` was necessary so combined it with the main tsconfig file. I also updated the linting and fixed some of those errors.

maastrich committed:
chore: Update @swc/types npm dependency to version 0.1.8

maastrich committed:
feat: Update AggregatedNode type to include ExportSpecifier and ImportSpecifier

maastrich committed:
feat: Fix typo in test/utils.ts

The commit message suggests that a typo has been fixed in the `test/utils.ts` file.

maastrich commented:
@morganney I can revert the meta tsconfig but I usually do that to ensure separation between the actual code and testing/config files
This way, you can ensure that the walker will run on the browser, no `node:*` needed but they will be defined for the tests 🤔 

Your choice

morganney commented:
> @morganney I can revert the meta tsconfig but I usually do that to ensure separation between the actual code and testing/config files This way, you can ensure that the walker will run on the browser, no `node:*` needed but they will be defined for the tests 🤔
> 
> Your choice

Gotcha. Thanks for explaining your reasoning.

I think in this case, for now anyway, can you combine it to just the one `tsconfig.json` file? If that ends up being a mistake or the other seems necessary, we can add it back in.

maastrich committed:
chore: Update tsconfig.json to include "test" and "eslint.config.js" in the "include" field

maastrich committed:
fix

maastrich commented:
@morganney sure, done

maastrich commented:
btw I've found few examples in swc main repository for the nodes I couldn't test, few of them are still untested but there aren't that much
about coverage, I believe some types aren't correct from swc as the decorators are never undefined 🤔

maastrich committed:
chore: Update npm dependency to latest stable version

maastrich committed:
chore: Fix unnecessary condition in TsPropertySignature method

The code changes in `baseVisitor.ts` fix an unnecessary condition in the `TsPropertySignature` method. The `for` loop now correctly handles the case when `n.params` is undefined. This change improves the code's readability and ensures proper execution.

morganney reviewed (approved):
@maastrich great work!

morganney commented:
I can tidy up the README in another PR and then release this later today.
